rpc: fundrawtransaction and walletcreatefundedpsbt also lock manually selected coins

Sjors opened the pull request:
When using `fundrawtransaction` and `walletcreatefundedpsbt` with `lockUnspents`, it would only lock automatically selected coins, not manually selected coins. That doesn't make much sense to me if the goal is to prevent accidentally double-spending yourself before you broadcast a transaction.

Note that when  creating a transaction, manually selected coins are automatic "unlocked" (or more accurately: the lock is ignored). Earlier versions of this PR introduced an error when a locked coin is manually selected, but this idea was abandoned after some discussion. An application that uses this RPC should either rely on automatic coin selection (with `lockUnspents`) or handle lock concurrency itself with manual coin selection. In particular it needs to make sure to avoid/pause calls with automatic coin selection between calling `lockunspent` and the subsequent spending RPC.    

See #7518 for historical background.

promag commented:
Concept ACK!

This made me think that we should fail funding if some coin is already locked?

promag commented on test/functional/rpc_psbt.py:
```diff
@@ -89,6 +89,15 @@ def run_test(self):
+        # Manually selected inputs can be locked:
+        utxo1 = self.nodes[0].listunspent()[0]
+        psbtx1 = self.nodes[0].walletcreatefundedpsbt([{"txid": utxo1['txid'], "vout": utxo1['vout']}], {self.nodes[2].getnewaddress():1}, 0,{"lockUnspents": True})["psbt"]
+        assert_equal(len(self.nodes[0].listlockunspent()), 1)
```
`assert_equal(len(self.nodes[0].listlockunspent()), 0)` before `walletcreatefundedpsbt`?

promag reviewed (commented):
Add a release note maybe? Clients unaware of this change can have dangling utxos locked.

Sjors commented:
Added a release note.

> This made me think that we should fail funding if some coin is already locked?

Maybe in a separate PR. I also think it's safer for users to explicitly unlock coins before using them.

promag commented:
> I also think it's safer for users to explicitly unlock coins before using them.

The unspent locking was done to deal with concurrent clients and so a client shouldn't take the decision to unlock a coin so lightly - it might screw with a concurrent funding. This is a possible use case and for this reason the user should be able to lock all or nothing IMO.

Sjors commented:
>  the lock-an-already-locked-coin-and-ignore behavior

Not sure what you mean. This PR can potentially lock a coin that's already locked, but why is that a problem? This PR does not ignore new things.

promag commented:
The problem is that you might screw with a concurrent funding. When you lock a coin you should expect that it won't be used by others.

Sjors commented:
If a concurrent wallet consumer uses manual coin selection, I don't think this PR makes their problem worse. If it uses automatic automatic coin selection then this PR has no impact.

Sjors commented:
After [IRC discussion](http://www.erisian.com.au/bitcoin-core-dev/log-2020-03-06.html#l-288) I changed this PR to honour locks for manually selected coins.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2150,6 +2150,7 @@ static UniValue lockunspent(const JSONRPCRequest& request)
                 "Temporarily lock (unlock=false) or unlock (unlock=true) specified transaction outputs.\n"
                 "If no transaction outputs are specified when unlocking then all current locked transaction outputs are unlocked.\n"
                 "A locked transaction output will not be chosen by automatic coin selection, when spending bitcoins.\n"
+                "With manual coin selection a locked transaction output must unlocked before use."
```
"... must be ..."

promag commented on test/functional/wallet_basic.py:
```diff
@@ -136,11 +136,20 @@ def run_test(self):
         unspent_0 = self.nodes[1].listunspent()[0]
         self.nodes[1].lockunspent(False, [unspent_0])
         tx = self.nodes[1].createrawtransaction([unspent_0], { self.nodes[1].getnewaddress() : 1 })
-        tx = self.nodes[1].fundrawtransaction(tx)['hex']
```
Are you removing this case because it's used elsewhere?

Sjors replied:
The original test was for auto-unlocking a coin, which now produces an error, so I check for that.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2453,15 +2453,19 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
+    LOCK(cs_wallet);
+
     for (const CTxIn& txin : tx.vin) {
+        if(IsLockedCoin(txin.prevout.hash, txin.prevout.n)) {
```
The final frontier.

promag commented on test/functional/wallet_basic.py:
```diff
@@ -136,11 +136,20 @@ def run_test(self):
+        # fundrawtransaction can lock an input
+        self.nodes[1].lockunspent(True, [unspent_0])
+        assert_equal(len(self.nodes[1].listlockunspent()), 0)
+        tx = self.nodes[1].fundrawtransaction(tx,{"lockUnspents": True})['hex']
```
nit, add space if you happen to push again.

promag reviewed (commented):
Code review ACK 991a38ce9e9c19ec78139d2988b2abd7c9983519.

instagibbs commented:
concept ACK

meshcollider reviewed (commented):
utACK 991a38ce9e9c19ec78139d2988b2abd7c9983519

Sjors commented:
Rebased

achow101 commented:
ACK 74992fd8cbe6f409bd1dad6c064deb5428509472

promag reviewed (commented):
> This made me think that we should fail funding if some coin is already locked?

After reading past discussions and some more careful analysis I now think this is not correct.

With the current approach, a client C1 must unlock the unspents to manually use them. But right after `lockunspent true ...` some other client C2 can fund/send and and take C1 unspents, and C1 will error.

So I think the "is locked" check should be used only for automatic coin selection, not for those explicitly set.

This means that:
 a) the client handles concurrency elsewhere and doesn't use locked unspents
 b) the client locks unspents beforehand and then use them by funding transactions with those as inputs - and also sets `lockUnspents` to grab the new ones
 c) the user makes consecutive fundings with `lockUnspents` set.
 
In other words, we can assume that the client that funds a transaction with locked inputs is the one that locked them in the first place.

The c) case above is relevant - with the current approach it wouldn't be possible because between each funding iteration all inputs would have to be unlocked.

Lastly, the option `lockUnspents` when `true` should always lock all inputs - the caller will "own" all inputs (existing inputs whether locked or not and the new ones). This was the original change.

Sorry @Sjors for the misdirection 😞

jonatack reviewed (commented):
Almost-ACK, review/tested, modulo I need to think about https://github.com/bitcoin/bitcoin/pull/18244#pullrequestreview-430047339.

Sjors commented:
I'll await the above discussion. Will address nits if that leads to a change.

Sjors committed:
[rpc] fundrawtransaction, walletcreatefundedpsbt lock manually selected coins

Previously only automatically selected coins were locked when lockUnspents is set.
It now also locks selected coins.

Sjors commented:
Based on the above discussion I changed this PR again, it now does the following:
1. lock all inputs for a transaction, rather than just the automatically selected ones
2. unchanged behaviour:  ignore lock on manually coins (with some tests)

promag commented:
Thanks @Sjors will review again.

meshcollider reviewed (commented):
Code review ACK 6d1f51343cf11b07cd401fbd0c5bc3603e185a0e

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2047,6 +2047,7 @@ static UniValue lockunspent(const JSONRPCRequest& request)
                 "Temporarily lock (unlock=false) or unlock (unlock=true) specified transaction outputs.\n"
                 "If no transaction outputs are specified when unlocking then all current locked transaction outputs are unlocked.\n"
                 "A locked transaction output will not be chosen by automatic coin selection, when spending bitcoins.\n"
+                "Manually selected coins are automatically unlocked.\n"
```
This is incorrect, or am I missing something?

Sjors replied:
Yes, but that's not the point. Automatically selection avoids locks, manually selection causes an unlock. This remark projects a user who assumes locked coins can't be accidentally spend by manually picking them.

fjahr replied:
I think "Locks on manually selected coins are ignored." would be a more intuitive description of what happens but I understand this to mean the same thing.

promag reviewed (commented):
Just want to point out that it's impossible to prevent 2 clients spend the same manually specified inputs.

One way to fix this is to make each client explicitly lock those inputs, and only proceed if the lock was granted.

Sjors commented:
Client-specific locks sounds fun, but not in the PR :-)

fjahr reviewed (commented):
Code review ACK 6d1f51343cf11b07cd401fbd0c5bc3603e185a0e
